File: day1/main.js
const fs = require("fs").promises

const part1 = (line) => {
	return line.split("")
		.filter(c => parseInt(c) == c)
		.map(c => parseInt(c));
}

const words = ["0", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
const regexp = new RegExp("^(" + words.join("|") + ")")

const part2 = (line) => {
	const out = []

	for(let i = 0; i < line.length; i++){
		if(parseInt(line[i]) == line[i]){
			out.push(parseInt(line[i]))
		}else{
			const match = line.substr(i).match(regexp)
			if(match){
				out.push(words.indexOf(match[0]))
			}
		}
	}

	return out
}

const compute = (input, extractor) => {
	return input.split("\n").map(i => {
		let z = extractor(i)
		return (z[0] * 10) + z[z.length - 1];
	})
	.reduce((a, b) => a + b, 0)

}

const read = async(file) => (await fs.readFile(file, "utf8")).trim()

const main = async () => {
	const test1 = await read("inputs/test1.txt")
	const test2 = await read("inputs/test2.txt")
	const input = await read("inputs/input.txt")
	console.log("Test 1", compute(test1, part1))
	console.log("Part 1", compute(input, part1))
	console.log("Test 2", compute(test2, part2))
	console.log("Part 2", compute(input, part2))
}
main()

File: day1/src/main.rs
use std::{fs, collections::HashMap};

fn main() {
  let test1 = fs::read_to_string("inputs/test1.txt").expect("ioerror");
  let test2 = fs::read_to_string("inputs/test2.txt").expect("ioerror");
  let input = fs::read_to_string("inputs/input.txt").expect("ioerror");

  let part1 = mapping_part1();
  let part2 = mapping_part2();

  compute(&test1, &part1);
  compute(&input, &part1);
  compute(&test2, &part2);
  compute(&input, &part2);
}

fn mapping_part1() -> HashMap<&'static str, i32>{
  let mut mapping: HashMap<&str, i32> = HashMap::new();
  mapping.insert("1", 1);
  mapping.insert("2", 2);
  mapping.insert("3", 3);
  mapping.insert("4", 4);
  mapping.insert("5", 5);
  mapping.insert("6", 6);
  mapping.insert("7", 7);
  mapping.insert("8", 8);
  mapping.insert("9", 9);
  mapping.insert("0", 0);
  return mapping;
}

fn mapping_part2() -> HashMap<&'static str, i32>{
  let mut mapping: HashMap<&str, i32> = HashMap::new();
  mapping.insert("1", 1);
  mapping.insert("2", 2);
  mapping.insert("3", 3);
  mapping.insert("4", 4);
  mapping.insert("5", 5);
  mapping.insert("6", 6);
  mapping.insert("7", 7);
  mapping.insert("8", 8);
  mapping.insert("9", 9);
  mapping.insert("0", 0);
  mapping.insert("one", 1);
  mapping.insert("two", 2);
  mapping.insert("three", 3);
  mapping.insert("four", 4);
  mapping.insert("five", 5);
  mapping.insert("six", 6);
  mapping.insert("seven", 7);
  mapping.insert("eight", 8);
  mapping.insert("nine", 9);
  return mapping;
}

fn compute(input :&String, map :&HashMap<&str, i32>){
  let data = input.trim();
  let mut sum = 0;
  for line in data.split("\n") {
    let mut vals: Vec<i32> = Vec::new();
    let mut i = 0;
    while i < line.len() {
      for keyw in map.keys() {
        if line[i..].starts_with(keyw) {
          vals.push(*map.get(keyw).unwrap());
          break;
        }
      }
      i += 1;
    }
    if !vals.is_empty(){
      sum += (vals.first().unwrap() * 10) + vals.last().unwrap();
    }
  }
  println!("Result: {}", sum);
}
